File: src/lib/inventoryPdf.ts
import { jsPDF } from "jspdf";
import { formatCurrency } from "./money";
import type { AppSettings, Product } from "./types";

/** Group in-stock products by category for the store checklist. */
export function groupInStockByCategory(products: Product[]): { category: string; items: Product[] }[] {
  const inStock = products.filter((p) => p.qty > 0);
  const map = new Map<string, Product[]>();
  for (const p of inStock) {
    const key = (p.category || "").trim() || "Uncategorized";
    const list = map.get(key) || [];
    list.push(p);
    map.set(key, list);
  }
  for (const list of map.values()) {
    list.sort((a, b) => a.name.localeCompare(b.name) || a.sku.localeCompare(b.sku));
  }
  const categories = [...map.keys()].sort((a, b) => {
    if (a === "Uncategorized") return 1;
    if (b === "Uncategorized") return -1;
    return a.localeCompare(b);
  });
  return categories.map((category) => ({ category, items: map.get(category)! }));
}

/**
 * Printable in-store inventory checklist — in-stock only, sorted by category.
 */
export function buildInventoryChecklistPdf(
  products: Product[],
  settings: AppSettings,
): jsPDF {
  const groups = groupInStockByCategory(products);
  const doc = new jsPDF({ unit: "mm", format: "letter" });
  const pageW = doc.internal.pageSize.getWidth();
  const pageH = doc.internal.pageSize.getHeight();
  const marginX = 14;
  const marginTop = 16;
  const marginBottom = 16;
  const usableW = pageW - marginX * 2;

  let y = marginTop;
  const biz = (settings.business_name || "My Business").trim();
  const today = new Date();
  const dateLabel = today.toLocaleDateString("en-US", {
    year: "numeric",
    month: "short",
    day: "numeric",
  });

  const ensureSpace = (need: number) => {
    if (y + need > pageH - marginBottom) {
      doc.addPage();
      y = marginTop;
    }
  };

  doc.setFont("helvetica", "bold");
  doc.setFontSize(16);
  doc.text(biz, marginX, y);
  y += 7;
  doc.setFontSize(12);
  doc.text("Inventory checklist — in stock", marginX, y);
  y += 5;
  doc.setFont("helvetica", "normal");
  doc.setFontSize(9);
  doc.setTextColor(80);
  doc.text(`Printed ${dateLabel} · qty greater than 0 · sorted by category`, marginX, y);
  doc.setTextColor(0);
  y += 8;

  const colSku = marginX;
  const colName = marginX + 28;
  const colQty = marginX + usableW - 38;
  const colPrice = marginX + usableW - 8;
  const colCheck = marginX + usableW;

  if (!groups.length) {
    doc.setFont("helvetica", "normal");
    doc.setFontSize(11);
    doc.text("No in-stock products.", marginX, y);
    return doc;
  }

  for (const group of groups) {
    ensureSpace(16);
    doc.setFillColor(245, 245, 245);
    doc.rect(marginX, y - 4, usableW, 7, "F");
    doc.setFont("helvetica", "bold");
    doc.setFontSize(11);
    doc.text(group.category, marginX + 1, y);
    doc.setFont("helvetica", "normal");
    doc.setFontSize(9);
    doc.text(`${group.items.length} item${group.items.length === 1 ? "" : "s"}`, colPrice, y, {
      align: "right",
    });
    y += 7;

    doc.setFont("helvetica", "bold");
    doc.setFontSize(8);
    doc.setTextColor(100);
    doc.text("SKU", colSku, y);
    doc.text("Name / details", colName, y);
    doc.text("Qty", colQty, y, { align: "right" });
    doc.text("Price", colPrice, y, { align: "right" });
    doc.text("✓", colCheck, y, { align: "right" });
    doc.setTextColor(0);
    y += 4;
    doc.setDrawColor(200);
    doc.line(marginX, y, marginX + usableW, y);
    y += 4;

    for (const item of group.items) {
      ensureSpace(10);
      doc.setFont("helvetica", "normal");
      doc.setFontSize(9);
      const sku = (item.sku || "—").slice(0, 14);
      doc.text(sku, colSku, y);
      const nameLine = item.name.slice(0, 42);
      doc.text(nameLine, colName, y);
      doc.text(String(item.qty), colQty, y, { align: "right" });
      doc.text(formatCurrency(item.price), colPrice, y, { align: "right" });
      // Checkbox for store count
      doc.setDrawColor(120);
      doc.rect(colCheck - 4, y - 3.2, 3.5, 3.5);
      y += 4.5;
      if (item.serial_number?.trim()) {
        ensureSpace(6);
        doc.setFontSize(8);
        doc.setTextColor(110);
        doc.text(item.serial_number.trim().slice(0, 70), colName, y);
        doc.setTextColor(0);
        y += 4;
      }
      y += 1.5;
    }
    y += 4;
  }

  ensureSpace(10);
  doc.setFont("helvetica", "normal");
  doc.setFontSize(8);
  doc.setTextColor(120);
  const totalItems = groups.reduce((n, g) => n + g.items.length, 0);
  const totalUnits = groups.reduce(
    (n, g) => n + g.items.reduce((s, p) => s + p.qty, 0),
    0,
  );
  doc.text(
    `${totalItems} products · ${totalUnits} units in stock`,
    marginX,
    y,
  );

  return doc;
}

export function downloadInventoryChecklistPdf(
  products: Product[],
  settings: AppSettings,
): void {
  const doc = buildInventoryChecklistPdf(products, settings);
  const stamp = new Date().toISOString().slice(0, 10);
  doc.save(`inventory-checklist-${stamp}.pdf`);
}

export function openInventoryChecklistPrint(
  products: Product[],
  settings: AppSettings,
): void {
  const doc = buildInventoryChecklistPdf(products, settings);
  const blob = doc.output("blob");
  const url = URL.createObjectURL(blob);
  const iframe = document.createElement("iframe");
  iframe.setAttribute("title", "Print inventory checklist");
  iframe.style.position = "fixed";
  iframe.style.right = "0";
  iframe.style.bottom = "0";
  iframe.style.width = "0";
  iframe.style.height = "0";
  iframe.style.border = "0";
  document.body.appendChild(iframe);
  iframe.onload = () => {
    try {
      iframe.contentWindow?.focus();
      iframe.contentWindow?.print();
    } finally {
      window.setTimeout(() => {
        iframe.remove();
        URL.revokeObjectURL(url);
      }, 60_000);
    }
  };
  iframe.src = url;
}

File: legacy/inventory_pdf.py
"""In-stock inventory checklist PDF for store floor counts."""

from __future__ import annotations

import io
from collections import defaultdict
from datetime import datetime

from reportlab.lib.pagesizes import letter
from reportlab.lib.units import inch
from reportlab.pdfgen import canvas

from database import get_all_settings, search_products
from models import Product
from utils import format_currency


def in_stock_products(products: list[Product] | None = None) -> list[Product]:
    items = products if products is not None else search_products("")
    return [p for p in items if p.qty > 0]


def group_by_category(products: list[Product]) -> list[tuple[str, list[Product]]]:
    groups: dict[str, list[Product]] = defaultdict(list)
    for p in products:
        key = (p.category or "").strip() or "Uncategorized"
        groups[key].append(p)
    for items in groups.values():
        items.sort(key=lambda p: ((p.name or "").lower(), (p.sku or "").lower()))

    def sort_key(name: str) -> tuple[int, str]:
        return (1 if name == "Uncategorized" else 0, name.lower())

    return [(name, groups[name]) for name in sorted(groups.keys(), key=sort_key)]


def build_inventory_checklist_pdf(
    products: list[Product] | None = None,
    settings: dict | None = None,
) -> bytes:
    """Letter-size checklist: in-stock only, sorted by category, with checkboxes."""
    if settings is None:
        try:
            settings = get_all_settings()
        except Exception:
            settings = {}
    biz = (settings.get("business_name") or "My Business").strip()
    in_stock = in_stock_products(products)
    grouped = group_by_category(in_stock)

    buffer = io.BytesIO()
    c = canvas.Canvas(buffer, pagesize=letter)
    width, height = letter
    margin_x = 0.6 * inch
    margin_top = 0.65 * inch
    margin_bottom = 0.6 * inch
    usable_w = width - 2 * margin_x

    y = height - margin_top
    today = datetime.now().strftime("%b %d, %Y")

    def ensure_space(need: float) -> None:
        nonlocal y
        if y - need < margin_bottom:
            c.showPage()
            y = height - margin_top

    c.setFont("Helvetica-Bold", 16)
    c.drawString(margin_x, y, biz)
    y -= 18
    c.setFont("Helvetica-Bold", 12)
    c.drawString(margin_x, y, "Inventory checklist — in stock")
    y -= 14
    c.setFont("Helvetica", 9)
    c.setFillColorRGB(0.35, 0.35, 0.35)
    c.drawString(
        margin_x,
        y,
        f"Printed {today}  ·  qty greater than 0  ·  sorted by category",
    )
    c.setFillColorRGB(0, 0, 0)
    y -= 20

    col_sku = margin_x
    col_name = margin_x + 1.1 * inch
    col_qty = margin_x + usable_w - 1.55 * inch
    col_price = margin_x + usable_w - 0.55 * inch
    col_check = margin_x + usable_w

    if not grouped:
        c.setFont("Helvetica", 11)
        c.drawString(margin_x, y, "No in-stock products.")
        c.save()
        return buffer.getvalue()

    for category, items in grouped:
        ensure_space(36)
        c.setFillColorRGB(0.94, 0.94, 0.94)
        c.rect(margin_x, y - 4, usable_w, 16, fill=1, stroke=0)
        c.setFillColorRGB(0, 0, 0)
        c.setFont("Helvetica-Bold", 11)
        c.drawString(margin_x + 4, y, category)
        c.setFont("Helvetica", 9)
        label = f"{len(items)} item" if len(items) == 1 else f"{len(items)} items"
        c.drawRightString(col_check, y, label)
        y -= 18

        c.setFont("Helvetica-Bold", 8)
        c.setFillColorRGB(0.4, 0.4, 0.4)
        c.drawString(col_sku, y, "SKU")
        c.drawString(col_name, y, "Name / details")
        c.drawRightString(col_qty, y, "Qty")
        c.drawRightString(col_price, y, "Price")
        c.drawRightString(col_check, y, "✓")
        c.setFillColorRGB(0, 0, 0)
        y -= 6
        c.setStrokeColorRGB(0.78, 0.78, 0.78)
        c.line(margin_x, y, margin_x + usable_w, y)
        y -= 12

        for item in items:
            ensure_space(28)
            c.setFont("Helvetica", 9)
            c.drawString(col_sku, y, (item.sku or "—")[:16])
            c.drawString(col_name, y, (item.name or "")[:48])
            c.drawRightString(col_qty, y, str(item.qty))
            c.drawRightString(col_price, y, format_currency(item.price))
            c.setStrokeColorRGB(0.45, 0.45, 0.45)
            c.rect(col_check - 10, y - 2, 9, 9, fill=0, stroke=1)
            y -= 12
            if (item.serial_number or "").strip():
                ensure_space(14)
                c.setFont("Helvetica", 8)
                c.setFillColorRGB(0.4, 0.4, 0.4)
                c.drawString(col_name, y, item.serial_number.strip()[:70])
                c.setFillColorRGB(0, 0, 0)
                y -= 10
            y -= 4

        y -= 8

    ensure_space(20)
    total_items = sum(len(items) for _, items in grouped)
    total_units = sum(p.qty for _, items in grouped for p in items)
    c.setFont("Helvetica", 8)
    c.setFillColorRGB(0.4, 0.4, 0.4)
    c.drawString(
        margin_x,
        y,
        f"{total_items} products  ·  {total_units} units in stock",
    )
    c.save()
    return buffer.getvalue()
